core/src/main/scala/org/apache/spark/ui/scope/RDDOperationGraph.scala

```diff
@@ -119,18 +119,47 @@ private[ui] object RDDOperationGraph extends Logging {
       { if (stage.attemptId == 0) "" else s" (attempt ${stage.attemptId})" }
     val rootCluster = new RDDOperationCluster(stageClusterId, stageClusterName)
 
+    var rootNodeCount = 0
+    val addRDDIds = new mutable.HashSet[Int]()
+    val dropRDDIds = new mutable.HashSet[Int]()
+
+    def isAllowed(ids: mutable.HashSet[Int], rdd: RDDInfo): Boolean = {
+      val parentIds = rdd.parentIds
+      if (parentIds.size == 0) {
+        rootNodeCount < retainedNodes
+      } else {
+        if (ids.size > 0) {
```

Comment:
The whole if-else is just `ids.isEmpty || parentIds.exists(id => ids.contains(id) || !dropRDDIds.contains(id))` ?